File: app/src/main/java/com/routesme/taxi_screen/java/Tracking/Class/TrackingHandler.java
package com.routesme.taxi_screen.java.Tracking.Class;

import android.content.Context;
import android.location.Location;
import android.util.Log;

import com.routesme.taxi_screen.java.Tracking.database.AppDatabase;
import com.routesme.taxi_screen.java.Tracking.database.AppExecutors;
import com.routesme.taxi_screen.java.Tracking.database.TrackingDao;
import com.routesme.taxi_screen.java.Tracking.model.Tracking;
import com.routesme.taxi_screen.java.Tracking.model.TrackingLocation;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.List;

import tech.gusavila92.websocketclient.WebSocketClient;

public class TrackingHandler {
    private static final String TAG = "TrackingHandler";
    private WebSocketClient trackingWebSocket;
    private SimpleDateFormat dateformat;
    private String currentStamptime;
    private AppDatabase mDb;
    private TrackingDao trackingDao;
    private Tracking tracking;
    private TrackingLocation trackingFirstLocation, trackingLastLocation;
    private Location firstLocation, lastLocation;


    public TrackingHandler() {
    }

    public TrackingHandler(Context context, WebSocketClient trackingWebSocket) {
        this.trackingWebSocket = trackingWebSocket;
        dateformat = new SimpleDateFormat("dd-MMM-yyyy hh:mm:ss aa");
        //Tracking ... Room Database...
        mDb = AppDatabase.getInstance(context.getApplicationContext());
        trackingDao = mDb.trackingDao();

    }


    public void sendOfflineTrackingToServer() {
        AppExecutors.getInstance().diskIO().execute(new Runnable() {
            @Override
            public void run() {
                if (trackingDao.loadAllLocations() != null) {
                    final List<Tracking> trackings = trackingDao.loadAllLocations();
                    if (!trackings.isEmpty()) {
                        for (int l = 0; l < trackings.size(); l++) {
                            tracking = trackings.get(l);
                            sendLocationViaSocket(tracking);
                        }
                        //clear Tracking Table....
                        clearTrackingTable();
                    } else {
                        Log.d(TAG, "trackingWebSocket ... no offline location is exists!");
                    }
                }
            }
        });
    }

    public void insertLocation(final TrackingLocation newLocation) {
        AppExecutors.getInstance().diskIO().execute(new Runnable() {
            @Override
            public void run() {
                currentStamptime = dateformat.format(new Date());
                final Tracking tracking = new Tracking(newLocation, currentStamptime);
                final List<Tracking> trackings = trackingDao.loadAllLocations();
                if (!trackings.isEmpty()) {
                    TrackingLocation lastTabletLocation = trackingDao.loadLastLocation().getLocation();
                    if (!newLocation.getLatitude().equals(lastTabletLocation.getLatitude()) && !newLocation.getLongitude().equals(lastTabletLocation.getLongitude())) {
                        trackingDao.insertLocation(tracking);
                    }
                } else {
                    trackingDao.insertLocation(tracking);
                }
            }
        });
    }


    public void locationChecker() {
        AppExecutors.getInstance().diskIO().execute(new Runnable() {
            @Override
            public void run() {
                if (trackingDao.loadAllLocations() != null) {
                    final List<Tracking> trackings = trackingDao.loadAllLocations();
                    if (!trackings.isEmpty()) {
                        trackingFirstLocation = new TrackingLocation(trackingDao.loadFirstLocation().getLocation().getLatitude(), trackingDao.loadFirstLocation().getLocation().getLongitude());
                        trackingLastLocation = new TrackingLocation(trackingDao.loadLastLocation().getLocation().getLatitude(), trackingDao.loadLastLocation().getLocation().getLongitude());
                        if (!trackingFirstLocation.getLatitude().equals(trackingLastLocation.getLatitude()) && !trackingFirstLocation.getLongitude().equals(trackingLastLocation.getLongitude())) {
                            float distance = getDistance();
                            if (distance >= 2) {
                                sendLocationViaSocket(trackingDao.loadLastLocation());
                                clearTrackingTable();
                            }
                        }

                    }
                }
            }
        });
    }

    private float getDistance() {
        firstLocation = new Location("First Location");
        lastLocation = new Location("Last Location");
        firstLocation.setLatitude(trackingFirstLocation.getLatitude());
        firstLocation.setLongitude(trackingFirstLocation.getLongitude());
        lastLocation.setLatitude(trackingLastLocation.getLatitude());
        lastLocation.setLongitude(trackingLastLocation.getLongitude());

        //Calculate distance between two locations...
        //Distance in Meter
        float distance = firstLocation.distanceTo(lastLocation);
        return distance;
    }



    private void sendLocationViaSocket(Tracking tracking) {
        String timestamp = tracking.getTimestamp();
        TrackingLocation trackingLocation = tracking.getLocation();
        String message = "location:" + trackingLocation.getLatitude() + "," + trackingLocation.getLongitude() + ";timestamp:" + timestamp;
        trackingWebSocket.send(message);
        Log.i("trackingWebSocket:  ", "Send message:  " + message);
    }

    public void clearTrackingTable() {
        AppExecutors.getInstance().diskIO().execute(new Runnable() {
            @Override
            public void run() {
                trackingDao.clearTrackingData();
            }
        });
    }


}

File: app/src/main/java/com/routesme/taxi_screen/kotlin/LocationTrackingService/Class/TrackingHandler.kt
package com.routesme.taxi_screen.kotlin.LocationTrackingService.Class

import android.annotation.SuppressLint
import android.content.Context
import android.location.Location
import android.util.Log
import com.routesme.taxi_screen.kotlin.LocationTrackingService.Database.AppDatabase
import com.routesme.taxi_screen.kotlin.Model.Tracking
import com.routesme.taxi_screen.kotlin.Model.TrackingLocation
import tech.gusavila92.websocketclient.WebSocketClient
import java.text.SimpleDateFormat
import java.util.*

class TrackingHandler(val context: Context, private var trackingWebSocket: WebSocketClient) {
    private val db = AppDatabase(context).trackingDao()

    @SuppressLint("SimpleDateFormat")
    private val dateFormat = SimpleDateFormat("dd-MMM-yyyy hh:mm:ss aa")

    fun sendOfflineTrackingLocationsToServer() {
        if (!db.loadAllLocations().isNullOrEmpty()) {
            for (t in db.loadAllLocations()) {
                sendLocationViaSocket(t)
            }
            clearTrackingTable()
        } else {
            Log.d("trackingWebSocketKotlin", "No offline location is exists!")
        }
    }

    fun insertLocation(trackingLocation: TrackingLocation) {
        val tracking = Tracking(0, trackingLocation, dateFormat.format(Date()))
        if (db.loadAllLocations().isNullOrEmpty()) {
            db.insertLocation(tracking)
            Log.d("trackingWebSocketKotlin", "Insert first location:  $tracking")
        } else {
            if (tracking.location != db.loadLastLocation().location) {
                db.insertLocation(tracking)
                Log.d("trackingWebSocketKotlin", "Insert new location:  $tracking")
            }
        }
    }

    fun locationChecker() {
        if (!db.loadAllLocations().isNullOrEmpty()) {
            val firstLocation = db.loadFirstLocation().location
            val lastLocation = db.loadLastLocation().location
            if (firstLocation != lastLocation) {
                if (distance(firstLocation, lastLocation) >= 2) {
                    sendLocationViaSocket(db.loadLastLocation())
                    clearTrackingTable()
                }
            }
        }
    }

    private fun distance(firstLocation: TrackingLocation, lastLocation: TrackingLocation): Float {
        return convertLocation(firstLocation,"firstLocation").distanceTo(convertLocation(lastLocation,"lastLocation"))
    }

    private fun convertLocation(trackingLocation: TrackingLocation, providerName:String):Location{
        val location = Location(providerName)
        location.latitude = trackingLocation.latitude
        location.longitude = trackingLocation.longitude
        return location
    }

    private fun clearTrackingTable() {
        db.clearTrackingData()
        Log.d("trackingWebSocketKotlin", "Clear tracking table!")
    }

    private fun sendLocationViaSocket(t: Tracking) {
        val location = "location:${t.location.latitude},${t.location.longitude};timestamp:${t.timestamp}"
        trackingWebSocket.send(location)
        Log.d("trackingWebSocketKotlin", "Send location : $location")
    }
}